contracts/SponsorFaucet.sol

```diff
@@ -64,38 +109,180 @@ contract SponsorFaucet is Ownable, Pausable, ReentrancyGuard {
     function isAppliable(address dapp) public returns (bool) {
         if (_isAppliableForGas(dapp) || _isAppliableForCollateral(dapp))
             return true;
+        require(dapp.isContract(), "ERROR_ADDRESS_IS_NOT_CONTRACT");
         _validateApplyForGas(dapp);
         _validateApplyForCollateral(dapp);
     }
 
+    //accept sponsor's cfx
+    function() external payable {}
+
+    //withdraw to specific address by amount
+    function withdraw(address payable sponsor, uint256 amount)
+        public
+        onlyOwner
+        nonReentrant
+        whenPaused
+    {
+        require(address(this).balance >= amount, "amount too high");
+        (bool success, ) = sponsor.call.value(amount)("");
+        require(success, "withdraw failed");
+    }
+
+    //set bounds for sponsorship
+    function setBounds(
+        uint256 gasTotalLimit,
+        uint256 collateralTotalLimit,
+        uint256 gasBound,
+        uint256 collateralBound,
+        uint256 upperBound
```

Comment:
Use `bounds` struct instead?

The same to all `setXxxBounds` methods

Reply:
modified.